Improve error message of functions related to `GetXlaTensor()`.

ysiraichi opened the pull request:
This PR refactors functions related to `GetXlaTensor()`, improving error message, and propagating status. These changes should make errors more actionable from users perspective. Additionally, this PR also allows Python API bound functions to returns status values. 

**Key Changes:**

- Add `GetValueOrThrow(Status)` overload, so that we can call `GetValueOrThrow` for both `Status` ans `StatusOr<T>` return values
- Properly handle status return types when binding Python functions, i.e. allow calls like`scope.def(..., [](...) -> absl::Status { ... })`
    - Calls `GetValueOrThrow()` in the resulting status, returning void
    - Lambda source information correctly appear in the status propagation trace
- Modify error messages of:
    - `GetXlaTensorImpl()`: assumes less context, with a more detailed error message
    - `GetXlaTensor()`: more meaningful, and higher level message
    - `GetXlaTensors()`: more information, such as the index of the non-XLA tensor
    - `ReplaceXlaTensor()`: reworded
- Create `GetInputXlaTensor()` for returning errors with more context
    - This should be used instead of calling `GetXlaTensor()`, specifically for inputs that directly come from the user
- Refactor 2 Python API implementation, as example:
    - `_xla_dot_general`: calling `GetInputXlaTensor()` on its tensor inputs
    - `_get_graph_hash`: calling `GetXlaTensors()` instead of an explicit loop with the same semantic

**Update (Aug 4):**

- Created `BuildStatusErrorMessage(const absl::Status&)`, in order to get the error message, and status propagation trace when crashing on a `ABSL_CHECK()` call.
    - Leaving the C++ stacktrace for future work
- Removed test for `_get_graph_hash`, since it crashes, now
- Added better error messages for `ReplaceTensor` functions

zhanyong-wan commented on test/dynamo/test_dynamo_integrations_util.py:
```diff
@@ -84,6 +84,18 @@ def test_get_graph_hash(self):
+    error_message = (
+        "Expected all tensors in the given list to be XLA tensors. "
+        "Element at index 1 is not an XLA tensor. Got: torch.FloatTensor")
+    with self.assertRaisesRegex(RuntimeError, expected_regex=error_message):
```
Can we do a straight string comparison instead of regex match? We don't want `.` to be interpreted as a regex meta character, for example. Also, I suspect that assertRaisesRegex() does a substring match (if it's like assertRegex()), while we need a full string match to ensure that we don't get undesired prefixes or suffixes.

ysiraichi replied:
Do you mean:

```python
try:
    # run actual test
except Exception as e:
    self.assertEqual(str(e), error_message)
```

I would argue that using `with assertRaisesRegex()` is better because:

1. It just looks better (e.g. easier to read, I know what to expect)
2. The `.` are only present without wildcards, and in the middle of the message. So, I thought that wouldn't be much of a problem. Let me know if you'd rather have me escape them.

zhanyong-wan commented on test/quantized_ops/test_dot_general.py:
```diff
@@ -56,6 +56,13 @@ def test_dot_general_int32_dtype(self):
+    lhs = torch.rand(10, 3, 4, dtype=torch.bfloat16)
+    rhs = torch.rand(10, 4, 5, dtype=torch.bfloat16, device="xla")
+    error_message = r"Expected input tensor `lhs` to be an actual XLA tensor. Got: CPUBFloat16Type"
+    with self.assertRaisesRegex(RuntimeError, expected_regex=error_message):
```
Ditto.

zhanyong-wan commented on torch_xla/csrc/aten_xla_bridge.h:
```diff
@@ -59,6 +59,15 @@ absl::StatusOr<absl_nonnull XLATensorPtr> GetXlaTensor(
 absl::StatusOr<std::vector<absl_nonnull XLATensorPtr>> GetXlaTensors(
     const at::ITensorListRef& tensors);
 
+// Retrieve the underlying `XLATensorPtr` from `tensor`.
```
Style: "Retrieves"

zhanyong-wan commented on torch_xla/csrc/aten_xla_bridge.h:
```diff
@@ -59,6 +59,15 @@ absl::StatusOr<absl_nonnull XLATensorPtr> GetXlaTensor(
+// functions, i.e. functions where the parameter name makes sense for
+// the end user.
+absl::StatusOr<absl_nonnull XLATensorPtr> GetInputXlaTensor(
+    const at::Tensor& tensor, const std::string_view param);
```
Style nit: the second const is ignored by the compiler and just adds clutter. Drop it?

zhanyong-wan commented on torch_xla/csrc/aten_xla_bridge.cpp:
```diff
@@ -80,8 +80,12 @@ static absl::StatusOr<XLATensorImpl * absl_nonnull> GetXlaTensorImpl(
-    return XLA_ERROR_WITH_LOCATION(absl::InvalidArgumentError(absl::StrCat(
-        "Input tensor is not an XLA tensor: ", tensor.toString())));
+    auto error_message =
+        absl::StrCat("Failed retrieving the inner XLATensorImpl* from ",
```
This error talks about C++ implementation details and isn't very actionable for users. How can a user trigger this error? Does it indicate a torch_xla bug? If so, we should crash instead.

ysiraichi replied:
`GetXlaTensorImpl()` function is actually an internal function. My reasoning, here, was that: since it is an internal function, it should have somewhat more detailed information. Callers to this function (e.g. `GetXlaTensor()`, `GetInputXlaTensor()`) should overwrite the error message with something more appropriate to the end user.

> How can a user trigger this error?

Usually, when the underlying `TensorImpl` of a tensor (after unwrapping the functional layer) is not a `XLATensorImpl` instance. In other words, when the argument is not a proper XLA tensor.

> Does it indicate a torch_xla bug?

No. In this case, it indicates that this function failed.

zhanyong-wan commented on test/dynamo/test_dynamo_integrations_util.py:
```diff
@@ -84,6 +84,18 @@ def test_get_graph_hash(self):
+        "Expected all tensors in the given list to be XLA tensors. "
+        "Element at index 1 is not an XLA tensor. Got: torch.FloatTensor")
+    with self.assertRaisesRegex(RuntimeError, expected_regex=error_message):
+      torch_xla._XLAC._get_graph_hash(tensors)
```
This is an internal API. Is it possible for a user to trigger this error via the public torch_xla API? If so, can we switch the test to use the public API so that it's more realistic?

zhanyong-wan commented on test/quantized_ops/test_dot_general.py:
```diff
@@ -56,6 +56,13 @@ def test_dot_general_int32_dtype(self):
+  def test_raises_error_on_non_xla_tensor(self):
+    lhs = torch.rand(10, 3, 4, dtype=torch.bfloat16)
+    rhs = torch.rand(10, 4, 5, dtype=torch.bfloat16, device="xla")
+    error_message = r"Expected input tensor `lhs` to be an actual XLA tensor. Got: CPUBFloat16Type"
```
The error message shouldn't reference the name `lhs` as in the actual user code the parameter name may be different. Just say "the first argument".

ysiraichi replied:
My reasoning for using `lhs` was that it's a formal parameter for `_xla_dot_general()`. When we call `help()` on it, or when it's called with mismatching parameter types, `lhs` and `rhs` names are there:

```python
>>> torch_xla._XLAC._xla_dot_general(5, 3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: _xla_dot_general(): incompatible function arguments. The following argument types are supported:
    1. (lhs: torch.Tensor, rhs: torch.Tensor, dimension_numbers: tuple, precision_config: Optional[str] = None, preferred_element_type: Optional[object] = None) -> torch.Tensor

Invoked with: 5, 3
```

Do you think we should replace it by "the first argument"?

zhanyong-wan replied:
I would prefer "the first argument" as it's clear by itself. If we write "lhs", the user would need to look up the documentation/help for `_xla_dot_general` to know what we are talking about.

zhanyong-wan commented on torch_xla/csrc/aten_xla_bridge.cpp:
```diff
@@ -99,21 +103,39 @@ absl::StatusOr<absl_nonnull XLATensorPtr> GetXlaTensor(
+      XLATensorPtr ptr, GetXlaTensor(tensor),
+      absl::StrCat("Expected input tensor `", param,
+                   "` to be an actual XLA tensor. Got: ", tensor.toString(),
+                   ". Consider moving `", param, "` to XLA before."));
```
Remove "before".

zhanyong-wan commented on torch_xla/csrc/aten_xla_bridge.cpp:
```diff
@@ -128,9 +150,11 @@ absl::Status ReplaceXlaTensor(const at::Tensor& tensor,
-                     tensors.size(), " vs. ", new_xla_tensors.size())));
+    std::string error_message = absl::StrCat(
+        "Expected the size of the list of tensors (", tensors.size(),
+        ") to match the size of the list of XLATensorPtr (",
```
This error message reference C++ details. If this indicate a torch_xla bug, we should crash here.

zhanyong-wan commented on torch_xla/csrc/init_python_bindings.cpp:
```diff
@@ -3336,19 +3368,17 @@ void InitXlaModuleBindings(py::module m) {
-          })
+      .def("_get_graph_hash",
+           [](const std::vector<at::Tensor>& tensors)
+               -> absl::StatusOr<py::bytes> {
```
I'm not sure about the motivation for allowing the callback to return a StatusOr. Since the callback is invoked by pytorch from python, there is no benefit for returning a StatusOr compared with TORCH_CHECK(). The biggest benefit seems to be that it lets us use XLA_ASSIGN_OR_RETURN in the callback, but we can use MaybeThrow instead.

ysiraichi replied:
I would argue that there are 2 main benefits:

1. As you mentioned, we get to call `XLA_ASSIGN_OR_RETURN()`
2. We get to overwrite the error message, and see the source location when `TORCH_SHOW_CPP_STACKTRACES` is set

While we could modify `MaybeThrow()` for doing (1), for (2) we would need to introduce another macro. Therefore, I think that calling `XLA_ASSIGN_OR_RETURN()` is better.

zhanyong-wan replied:
I still don't see the benefits:

- (1) is not a benefit as we can use MaybeThrow().
- for (2) we can just call TORCH_CHECK() directly.

zhanyong-wan reviewed (commented):
Thanks!

zhanyong-wan commented on torch_xla/csrc/init_python_bindings.cpp:
```diff
@@ -87,6 +88,26 @@ namespace {
+// Traits related to the return type of the lambda function that wraps the
+// actual implementation inside PythonScope.
+template <class T>
+struct LambdaReturnTypeTrait {
```
The name should make its semantics obvious. Since this is used to unwrap Status/StatusOr, call it `StatusTypeTrait`?

ysiraichi replied:
Semantically, I think `StatusTypeTrait` falls short, because its inner `Type` is also defined for non-status types:

- `StatusTypeTrait<T>::Type == T`
- `StatusTypeTrait<Status>::Type == void`
- `StatusTypeTrait<StatusOr<T>>::Type == T`

So, I named it `LambdaReturnTypeTrait` because it gives us the return type of the actual wrapping lambda, given the implementation function return type.

What do you think of: `WrapperLambdaReturn<ImplRetType>::type`?

zhanyong-wan commented on torch_xla/csrc/init_python_bindings.cpp:
```diff
@@ -87,6 +88,26 @@ namespace {
+// actual implementation inside PythonScope.
+template <class T>
+struct LambdaReturnTypeTrait {
+  using Type = T;
```
`Type` is unclear. Also for trait classes we typically follow the STL naming convention of lowercase members.

`raw_type`?

ysiraichi replied:
Right. `Type` actually represents the return type of the wrapper lambda, given `T` is the return type of the implementation function. So, as mentioned above, what do you think of: `WrapperLambdaReturn<ImplRetType>::type`?

zhanyong-wan replied:
If we call the class `RemoveStatus`, then it makes sense to call this `type`, as `RemoveStatus<T>::type` is clear.

ysiraichi commented:
Besides the changes requested in the review, other relevant changes are:

- Created `BuildStatusErrorMessage(const absl::Status&)`, in order to get the error message, and status propagation trace when crashing on a `ABSL_CHECK()` call.
    - Leaving the C++ stacktrace for future work
- Removed test for `_get_graph_hash`, since it crashes, now
- Added better error messages for `ReplaceTensor` functions

ysiraichi commented on torch_xla/csrc/status.h:
```diff
@@ -200,6 +211,9 @@ T GetValueOrThrow(absl::StatusOr<T>&& status) {
 }
 
+// `GetValueOrThrow` overload for `Status`.
+void GetValueOrThrow(const absl::Status& status);
```
In a way, the "value" (i.e. non-error value) inside `absl::Status` is nothing (i.e. void). So, I wouldn't say this is misleading. Also, `GetValueOrThrow` is supposed to "consume" the status type, returning its inner type. So, I would rather return void, here.

ghpvnist reviewed (approved):
LGTM pending @zhanyong-wan's approval.

zhanyong-wan reviewed (approved):
Thanks!
